### PR TITLE
Add a "high-speed" version of `autoReg`

### PR DESCRIPTION
**Pull request by christiaanb**
This still splits registers of products into products of registers, but it no longer uses the tag bit to control the enable lines of the registers holding the value bits. This decreases the logic depth, thus shortening the potential critical path, at the cost of power, since the fields bit can freely toggle.

### CODE REVIEW

File: clash-prelude/src/Clash/Class/AutoReg/Internal.hs [hunk 1]
```diff
@@ -113,6 +113,19 @@ class NFDataX a => AutoReg a where
   autoReg = register
   {-# INLINE autoReg #-}
 
+  -- | A "High-Speed" version of 'autoReg', that still splits product types
+  -- into multiple registers, but no longer uses the tag bit to control the
+  -- enable of the field bits. This reduces the potential logic depth, at the
+  -- cost of power due to toggling field bits.
+  autoRegHS
```
Comment (martijnbastiaan) [inline clash-prelude/src/Clash/Class/AutoReg/Internal.hs:120]: Drive-by nitpick: `autoRegHs` maybe? Makes it easier to extend (`autoRegHsFoo` vs `autoRegHSFoo`) and a bit easier to type.

Comment (christiaanb) [reply clash-prelude/src/Clash/Class/AutoReg/Internal.hs]: Done

File: clash-prelude/src/Clash/Class/AutoReg/Internal.hs [hunk 2]
```diff
@@ -160,6 +173,22 @@ instance AutoReg a => AutoReg (Maybe a) where
...
   {-# INLINE autoReg #-}
 
+  autoRegHs clk rst en initVal input = createMaybe <$> tagR <*> valR
+   where
+     tag = isJust <$> input
+     tagInit = isJust initVal
+     tagR = suffixName @"tag" register clk rst en tagInit tag
+
+     val = fromMaybe (deepErrorX "autoReg'.val") <$> input
+     valInit = fromMaybe (deepErrorX "autoReg'.valInit") initVal
```
Comment (leonschoorl) [inline clash-prelude/src/Clash/Class/AutoReg/Internal.hs:183]:
```suggestion
     val = fromMaybe (deepErrorX "autoRegHs.val") <$> input
     valInit = fromMaybe (deepErrorX "autoRegHs.valInit") initVal
```

Comment (leonschoorl) [reply clash-prelude/src/Clash/Class/AutoReg/Internal.hs:183]: And could you also remove the `'`s from the error strings in the original `autoReg`, which I accidentally left in there?

**Review by leonschoorl — CHANGES_REQUESTED** [review]

**Comment by leonschoorl** [comment]
Other then those two comments, LGTM

**Review by leonschoorl — CHANGES_REQUESTED** [review]
You should also add `autoRegHs` to the RTree instance and to the [TH code that generates instance for product types](https://github.com/clash-lang/clash-compiler/blob/b0a239e1a6b0e0d0f7cf63a0c85f2c1448a0a13e/clash-prelude/src/Clash/Class/AutoReg/Internal.hs#L336)

**Comment by christiaanb** [comment]
Yeah... need to fix that hard-to-fix bug in Clash first as well...
